File: python/Challenge_problem.py

# coding: utf-8

# In[17]:


import scipy.integrate as sci
import numpy as np
import math
from matplotlib import pyplot

N = 80
T1 = []
T2 = []
T3 = []
A = np.linspace(0.1,4,N)

for i in A:
    dT = lambda b:((2*2**0.5)/(i**4-b**4)**0.5)
    b = np.linspace(0,i,N,endpoint=False)
    dT1vec = np.vectorize(dT)
    dT1_b = dT1vec(b)
    T1.append(sci.trapz(dT1_b, b)) # calculate time period by Trapezoidal integration method
    T2.append(sci.quad(dT,0,i)[0]) # calculate time period by normal python integration
    T3.append(sci.simps(dT1_b, b)) # calculate time period by Simpson's rule
    
    
pyplot.plot(b,T1,'g*-', label ="Time period by Trapezoidal rule")
pyplot.plot(b,T2,'r--', label ="Time period by normal python integration rule")
pyplot.plot(b,T3,'b--', label ="Time period by Simpson's rule")
pyplot.xlabel('Amplitude')
pyplot.ylabel('Period')
pyplot.title("T vs. A by different integration methods")
pyplot.legend()

File: MATLAB/Challenge_problem.m
clear all

dt = @(x1,a) (2*sqrt(2)) ./sqrt(a .^4 - x1 .^4);
Amp = linspace(0.1,7,100);
T = [];
for i=1:length(Amp)
   Amp1 = Amp(i);
   x = linspace(0.01,Amp1*99/100);
   dT = trapz(x,dt(x,Amp1));
   T = [T, dT]
    
end

plot(Amp,T)
xlabel('Amplitude')
ylabel('Period')
title('Anharmonic oscillator')